integration/querier_remote_read_test.go

```diff
@@ -59,164 +63,258 @@ func TestQuerierRemoteRead(t *testing.T) {
+		query           *prompb.Query
+		expectedStartMs int64
+		expectedEndMs   int64
 	}{
-		"float samples": {
-			expectedValType: chunkenc.ValFloat,
-			floats: func(startMs, endMs int64) []prompb.Sample {
-				var samples []prompb.Sample
-				for i := startMs; i < endMs; i++ {
-					samples = append(samples, prompb.Sample{
-						Value:     rand.Float64(),
-						Timestamp: i,
-					})
-				}
-				return samples
+		"float samples, remote read request without hints": {
+			valType:    chunkenc.ValFloat,
+			metricName: floatMetricName,
+			query: &prompb.Query{
+				Matchers:         remoteReadQueryMatchersByMetricName(floatMetricName),
+				StartTimestampMs: pushedStartMs,
+				EndTimestampMs:   pushedEndMs,
 			},
+			expectedStartMs: pushedStartMs,
+			expectedEndMs:   pushedEndMs,
 		},
-		"histograms": {
-			expectedValType: chunkenc.ValHistogram,
-			histograms: func(startMs, endMs int64) []prompb.Histogram {
-				var hists []prompb.Histogram
-				for i := startMs; i < endMs; i++ {
-					h := test.GenerateTestHistogram(int(i))
-					hists = append(hists, remote.HistogramToHistogramProto(i, h))
-				}
-				return hists
+		"float samples, with hints time range equal to query time range": {
+			valType:    chunkenc.ValFloat,
+			metricName: floatMetricName,
+			query: &prompb.Query{
+				Matchers:         remoteReadQueryMatchersByMetricName(floatMetricName),
+				StartTimestampMs: pushedStartMs,
+				EndTimestampMs:   pushedEndMs,
+				Hints: &prompb.ReadHints{
+					StartMs: pushedStartMs,
+					EndMs:   pushedEndMs,
+				},
 			},
+			expectedStartMs: pushedStartMs,
+			expectedEndMs:   pushedEndMs,
 		},
-		"float histograms": {
-			expectedValType: chunkenc.ValFloatHistogram,
-			histograms: func(startMs, endMs int64) []prompb.Histogram {
-				var hists []prompb.Histogram
-				for i := startMs; i < endMs; i++ {
-					h := test.GenerateTestFloatHistogram(int(i))
-					hists = append(hists, remote.FloatHistogramToHistogramProto(i, h))
-				}
-				return hists
+		"float samples, with hints time range different than query time range": {
+			valType:    chunkenc.ValFloat,
+			metricName: floatMetricName,
+			query: &prompb.Query{
+				Matchers:         remoteReadQueryMatchersByMetricName(floatMetricName),
+				StartTimestampMs: pushedStartMs,
+				EndTimestampMs:   pushedEndMs,
+				Hints: &prompb.ReadHints{
+					StartMs: now.UnixMilli(),
+					EndMs:   now.Add(10 * time.Second).UnixMilli(),
+				},
 			},
+			// Mimir doesn't cut returned chunks to the requested time range for performance reasons. This means
+			// that we get the entire chunks in output. TSDB targets to cut chunks once every 120 samples, but
+			// it's based on estimation and math is not super accurate. That's why we get these not-perfectly-aligned
+			// chunk ranges. However, they're consistent, so tests are stable.
+			expectedStartMs: now.Add(-1500 * time.Millisecond).UnixMilli(),
+			expectedEndMs:   now.Add(10100 * time.Millisecond).UnixMilli(),
```

Comment:
Note to reviewers: this is due to how TSDB `computeChunkEndTime()` works. I've confirmed it doing the math manually.